src/entt/entity/polymorphic.hpp

```diff
@@ -0,0 +1,500 @@
+ */
+
+
+namespace entt {
+
+/** @brief base class for entt::poly_pool_holder */
+template<typename Entity>
+class poly_pool_holder_base {
+public:
+    inline poly_pool_holder_base(void* pool,
+                                 void* (*getter)(void*, Entity) ENTT_NOEXCEPT) :
+          pool_ptr(pool), getter_ptr(getter) {}
+
+protected:
+    void* pool_ptr;
+    void* (*getter_ptr)(void*, Entity) ENTT_NOEXCEPT;
+};
+
+/**
+ * @brief Holds pointer to child type storage and a function to convert it into parent type
+ * @tparam Entity underlying entity type
+ * @tparam Type polymorphic component type to convert into
+ * @tparam Allocator allocator type of the pool
+ */
+template<typename Entity, typename Type, typename Allocator>
+class poly_pool_holder : public poly_pool_holder_base<Entity> {
+public:
+    using poly_pool_holder_base<Entity>::poly_pool_holder_base;
+
+    /** @brief entity type of the component pools */
+    using entity_type = Entity;
+    /** @brief component type */
+    using value_type = Type;
+    /** @brief pointer to the component, will be same as value_type, if component itself is a pointer */
+    using pointer_type = std::remove_pointer_t<Type>*;
+    /** @copydoc pointer_type */
+    using const_pointer_type = const std::remove_pointer_t<Type>*;
+    /** @brief type of hold sparse set */
+    using sparse_set_type = basic_sparse_set<Entity, typename std::allocator_traits<Allocator>::template rebind_alloc<Entity>>;
+
+    /**
+     * @brief gets child from a child pool as parent type reference by given entity
+     * @param ptr pointer from the child type set
+     * @return Returns reference to Type, converted from a given pointer. Will return pointer instead of reference, if parent type is a pointer.
+     */
+    inline pointer_type try_get(const Entity ent) ENTT_NOEXCEPT {
+        return pointer_type(this->getter_ptr(this->pool_ptr, ent));
+    }
+
+    /** @copydoc try_get */
+    inline const_pointer_type try_get(const Entity ent) const ENTT_NOEXCEPT {
+        return const_cast<poly_pool_holder<Entity, Type, Allocator>*>(this)->try_get(ent);
+    }
+
+    /** @brief returns underlying pool */
+    inline auto& pool() ENTT_NOEXCEPT {
+        return *static_cast<sparse_set_type*>(this->pool_ptr);
+    }
+
+    /** @copydoc pool */
+    inline const auto& pool() const ENTT_NOEXCEPT {
+        return *static_cast<sparse_set_type*>(this->pool_ptr);
+    }
+
+    /**
+     * @brief removes component from pool by entity
+     * @param ent entity
+     * @return true, if component was removed, false, if it didnt exist
+     */
+    inline bool remove(const Entity ent) {
+        return pool().remove(ent);
+    }
+};
+
+/**
+ * @brief Holds runtime information about one polymorphic component type.
+ * @tparam Entity underlying entity type
+ * @tparam Type polymorphic component type
+*  @tparam Allocator allocator type for all bound contained pool holders
+ */
+template<typename Entity, typename Type, typename Allocator>
```

Comment:
This is similar to `poly_pool_holder` and I wonder why we don't get the storage type directly rather than a bunch of things that can quickly go out of sync.